### ISSUE TITLE
Mouse right click or taps not working

### ISSUE THREAD

**Issue by abhisek176** [MCMrARM/mbp2018-bridge-drv#10]
HI,
Thanks for the great help with the modules. Keyboard and touchpad works just fine with MBP15,4. However, the right click and taps on the touchpad is a problem. Please let me know if something can be done.

**Comment by yurnam** [comment]
Hi,  try using  bce together with hid-apple from https://github.com/MCMrARM/mbp2018-etc 

Works fine for me on my 15.4.

**Comment by abhisek176** [comment]
I did install both of them. But Audio,wifi and touchpad right click,scrolling and tap without clicks not working. Everything works for you? Can you tell me the Kernel Version or how do you ingest them?

**Comment by yurnam** [comment]
i use 5.5-rc4 with the config attached. 
also applied only the keyboard and wifi patches from https://github.com/aunali1/linux-mbp-arch

just copied the .ko files in /lib/modules/$(uname -r) and ran depmod
everything got loaded automatically on reboot.

Working:

- touchbar, 
- audio (Speaker only) 
- Keyboard
- Keyboard-backlight
- nvme
- Trackpad with right click, scroll and tap


Wifi and Bluetooth are currently not working, due to a new Broadcom chip (BCM4377) used in some of these models.
See:
https://github.com/Dunedan/mbp-2016-linux/issues/112









[config.txt](https://github.com/MCMrARM/mbp2018-bridge-drv/files/4017470/config.txt)

**Comment by abhisek176** [comment]
Thanks for the reply @yurnam . Could you please share an easy way to upgrade the kernel to 5.5-rc4 and if I can reinstall all of them at once (the available drivers)! That would be a great help. I am using Ubuntu 18.04

**Comment by yurnam** [comment]
what kernel version are you using right now ?
I'll write a tutorial tomorrow.  :)

**Comment by abhisek176** [comment]
I am using 5.0.0-37. That would be really awesome (waiting :+1: ). BTW I do not have to reinstall Ubuntu again, right (Just a noob's question :laughing: )?

**Comment by yurnam** [comment]
No you don't have to reinstall ubuntu 
step 1 just follow this guide here: https://www.cyberciti.biz/tips/compiling-linux-kernel-26.html
but make sure you do `make modules_install INSTALL_MOD_STRIP=1` otherwise it will hang at grub

**Comment by abhisek176** [comment]
This is great! Thanks! Is is needed to include @aunali1's patches here?

**Comment by yurnam** [comment]
honestly i am not sure yet, i'll try it

**Comment by abhisek176** [comment]
Alright then, I will go ahead and install the kernel along with the bce,hid-apple and touchbar drivers. Let me see if I can get things done now. Thanks a lot for your help. (Please do a tutorial on this as it would help a lot)

**Comment by abhisek176** [comment]
Hi @yurnam : Is it possible to edit these patches https://github.com/ppaulweber/linux-mba for 5.4.7. It seems all included. I am not able to append the .diff patches.

**Comment by abhisek176** [comment]
Tried with the latest kernel, still touchpad things are not working.. Rest are working fine .. @yurnam

**Comment by yurnam** [comment]
can you tell me exactly whats working and what's not ?

**Comment by abhisek176** [comment]
Working:

touchbar,
audio (Speaker only)
Keyboard
Keyboard-backlight
nvme
Trackpad with only right or left click

Not working:
Trackpad scroll, taps and both the clicks at once
Fn key seems not working

**Comment by yurnam** [comment]
https://github.com/yurnam/linux-mbp15-4 works for me

**Comment by abhisek176** [comment]
Have you complied everything into this?

**Comment by yurnam** [comment]
updated it a minute ago, bce and touchbar were added, but not compiling, now it's working w/o external patches, 

also: please check if you have two finger scrolling enabled in gnome-settings ;)

**Comment by abhisek176** [comment]
Gnome does not show any option for that.

**Comment by yurnam** [comment]
gnome-settings --> Devices --> Mouse and Touchpad --> Two - Finger Scrolling

might be a bit off because mine is in German :)

**Comment by abhisek176** [comment]
There are no option available : https://drive.google.com/file/d/1LZFESEISk3hY2-pVV0Ad_awqCmW3czpk/view?usp=drivesdk

**Comment by yurnam** [comment]
should look like this

![DeepinScreenshot_select-area_20200104182520](https://user-images.githubusercontent.com/25354628/71769298-95169c00-2f1f-11ea-8889-db15a249f24d.png)

**Comment by abhisek176** [comment]
right. :( what distro are you using. It does not look like Ubutnu!

**Comment by yurnam** [comment]
I use ubuntu 19.10 with a custom theme  

pls try downloading my repo and run the mk.sh script to install

**Comment by abhisek176** [comment]
Alright. Let me try that. Otherwise I will use Ubuntu 19.10

**Comment by yurnam** [comment]
19.10 required to blacklist the efivarfs module, otherwise it would crash on boot
did not happen in 18.04

**Comment by abhisek176** [comment]
*If this does not work that's the last option.. :(

**Comment by abhisek176** [comment]
[
[make.log](https://github.com/MCMrARM/mbp2018-bridge-drv/files/4022473/make.log)
](url)

**Comment by abhisek176** [comment]
problem compiling it.

**Comment by abhisek176** [comment]
Tried with Ubuntu 19.10. Same issue.. ;(

**Comment by yurnam** [comment]
okay seems like i forgot to push to remote

should work now

**Comment by abhisek176** [comment]
no same issue. I think the drivers is not a problem! There is something else. Can you tell me which package are you using to tweak the them. let me try installing that.

**Comment by yurnam** [comment]
okay, try disabling CONFIG_TOUCHBAR_DRIVER

**Comment by abhisek176** [comment]
Then it will be same as installing 5.5.rc4 kernel and then installing these additional t2 drivers which I did.

**Comment by yurnam** [comment]
the bce (sound, touchpad) driver is included and compiles
i tried to include the mbp15 branch from https://github.com/roadrunner2/macbook12-spi-driver
but this driver won't compile yet
just install this driver seperately

**Comment by abhisek176** [comment]
I installed 5.5.rc4 and them compiled bce+mbp15 (It's the same?!) .. The problem is same. I think there is something else happening which I can't figure out. Do I need synaptics separately ? Can you post your ls \etc\X11\  outputs

**Comment by abhisek176** [comment]
And btw can you tell me which package you installed for that gnome theme? Let me try that and see.

**Comment by yurnam** [comment]
so, fixed the compiling issue now everything compiles,
 i got a macOS theme for ubuntu from https://www.gnome-look.org/

**Comment by abhisek176** [comment]
Okay, I will try it for a final check. Do I need to remove the 5.5 rc4 and additional bce kernel before installing this?

**Comment by yurnam** [comment]
you don't have to remove them

**Comment by abhisek176** [comment]
Did not work that either.  Can you please do me a favor! Please share your .ko files (bce.ko, hid-apple.ko etc). I can put them directly and see if it works!

**Comment by abhisek176** [comment]
I am trying your new kernel. In the meanwhile can you send me your .ko files? And can you check
[usb.txt](https://github.com/MCMrARM/mbp2018-bridge-drv/files/4025905/usb.txt)
 the USB ids?

**Comment by yurnam** [comment]
these ko files wont work in your kernel you need to build them yourself

**Comment by yurnam** [comment]
https://github.com/yurnam/linux-mbp15-4/releases/tag/ubuntu-package 

i made a package for ubuntu

**Comment by abhisek176** [comment]
Ahh. That's great. Let me try it now. BTW 'make' failed again with same error.

**Comment by abhisek176** [comment]
Hi, the boot fails with your build package. It seems there is a .xz extension for every kernel module which may be causing the boot error.

For example it should be bce.ko inside the drivers folder /lib/module/uname-r/kernel/drivers.

**Comment by yurnam** [comment]
it boots fine on mine no boot errors and the kernel should support loading modules with xz compression

**Comment by abhisek176** [comment]
It does not for mine.. :( I am getting boot error with loading kernel errors. So I check with the default kernel and found this difference.

**Comment by yurnam** [comment]
what error are you getting?

**Comment by abhisek176** [comment]
loading kernel errors.

**Comment by yurnam** [comment]
can you get into recovery mode?

**Comment by abhisek176** [comment]
Yes. But then my external keyboard etc do not work.

**Comment by yurnam** [comment]
i made a mistake in the deb package... 
the file /lib/modprobe.d/blacklist-mbp15-4.conf should be 
/etc/modprobe.d/blacklist-mbp15-4.conf

ill fix that and upload a new package

**Comment by yurnam** [comment]
the modules 
efivarfs
brcmfmac

are causing crashes so they need to be blacklisted for now

**Comment by abhisek176** [comment]
Okay. Please let me know. See this https://drive.google.com/file/d/1M5kanT_1dJiJBOyXHRbjuUH74bCLddpb/view?usp=drivesdk

**Comment by yurnam** [comment]
i dont have access

**Comment by abhisek176** [comment]
my mistake. check it now.

**Comment by yurnam** [comment]
can you please do `$ ls /lib/modules` for me

**Comment by abhisek176** [comment]
[mod.txt](https://github.com/MCMrARM/mbp2018-bridge-drv/files/4026203/mod.txt)

**Comment by yurnam** [comment]
try to run `# depmod 5.5.0-rc4-yurnam-gd9eb59b6f`

**Comment by abhisek176** [comment]
unable to boot into the system and in recover mode keyboard not working.

**Comment by yurnam** [comment]
you can do this in any working kernel

**Comment by abhisek176** [comment]
okay. I did it , now?

**Comment by yurnam** [comment]
reboot into this kernel

**Comment by yurnam** [comment]
new package here: https://github.com/yurnam/linux-mbp15-4/releases/tag/ubuntu-new

**Comment by abhisek176** [comment]
Sorry, It shows the same error while booting.. :(

**Comment by abhisek176** [comment]
Is it possible to make the drivers with .ko extension?

**Comment by yurnam** [comment]
the problem is that the kernel doesn't know where the module dependencies are, so you need to somehow run depmod while in the kernel but if you dont get any keyboard working you could try to setup  a script that runs the command on startup before it tries to load the kernel modules
unfortunately i have no idea how to do this

**Comment by abhisek176** [comment]
Alright then. I will wait for some solution. Thanks for your help.. :)

**Comment by yurnam** [comment]
i'll try compiling on a different computer and see if it really just works on my mac

**Comment by abhisek176** [comment]
Okay. Please let me know here.

**Comment by yurnam** [comment]
did you compile with touchbar, bce and t2 as module or as builtin ?

**Comment by abhisek176** [comment]
With 5.5.rc5 or with your kernel?

**Comment by yurnam** [comment]
my kernel

**Comment by abhisek176** [comment]
I just run sudo dpkg -i ****.deb to install it.

**Comment by yurnam** [comment]
no i mean when you tried to compile from my source

**Comment by abhisek176** [comment]
I used builtin, your makefile as it is.

**Comment by yurnam** [comment]
please do a fresh git clone and then run mk.sh :)

**Comment by yurnam** [comment]
builtin produces the error the bce touchbar and t2 need to be modules not builtin
 
i added a .config into the repo with the right configuration

**Comment by abhisek176** [comment]
when I run it asks me to include bce and touchbar , I put yes. My point is if I have to install the module separately then I guess there will be no difference between original Kernel 5.5.rc4 and this ? Am I right?

**Comment by abhisek176** [comment]
What I am doing now is installing 5.5-rc4 and then installing bce and tocuhbar as modules. Won't it be same then?

**Comment by yurnam** [comment]
you should type M. not Y
no this is not the same 
Y compiles the driver into the vmlinuz file,  this is not supported for every driver

M compiles the driver seperately as a .ko file which can be loaded reloaded and unloaded

**Comment by abhisek176** [comment]
Yes correct.. My point is won't be same then as I am doing. Installing kernel first and then adding the modules separately?

**Comment by abhisek176** [comment]
for e.g. Now i install 5.5 rc5 kernel, and installing the bce and touchbar modules..?!!!

**Comment by yurnam** [comment]
that's why i created this repo, so that this is no longer needed :D

**Comment by abhisek176** [comment]
Correct. But this is same? no?

**Comment by yurnam** [comment]
I just added the custom drivers, that worked for me to the kernel so they don't have to be built seperately all you need to to is clone the repo, run the mk.sh script and wait :)

**Comment by abhisek176** [comment]
okay. I will see if this works.

**Comment by abhisek176** [comment]
Can you check the usb device configurations? I raised an issue in https://github.com/ppaulweber/linux-mba/issues/1#issuecomment-571122621 and it seems the device  id is still not in the upstream kernel. Is that a problem?

**Comment by yurnam** [comment]
I dont think that's a problem since i didn't even know that.
Everything usb related works for me

**Comment by abhisek176** [comment]
Fails instantly : https://drive.google.com/file/d/1MSOlU9pkmxMHK8FqeF3DHQlY0Y_91ynd/view?usp=drivesdk

**Comment by yurnam** [comment]
apt install libssl-dev libssl

**Comment by abhisek176** [comment]
I did that.  Okay let me try again
